src/dfx/src/commands/deps/pull.rs

```diff
@@ -0,0 +1,340 @@
+            error!(logger, "Failed to pull canister {canister_id}.\n{e}");
+            any_download_fail = true;
+        }
+    }
+
+    match any_download_fail {
+        true => bail!("Failed when pulling canisters."),
+        false => {
+            for (name, canister_id) in &pull_canisters_in_config {
+                copy_service_candid_to_project(&project_root, name, *canister_id)?;
+            }
+        }
+    }
+
+    save_pulled_json(&project_root, &pulled_json)?;
+    Ok(())
+}
+
+#[context("Failed to fetch and parse `dfx:deps` metadata from canister {canister_id}.")]
+async fn fetch_deps_to_pull(
+    agent: &Agent,
+    logger: &Logger,
+    canister_id: Principal,
+    canisters_to_pull: &mut VecDeque<Principal>,
+    pulled_json: &mut PulledJson,
+) -> DfxResult {
+    info!(
+        logger,
+        "Resolving dependencies of canister {canister_id}..."
+    );
+
+    match fetch_metatdata(agent, canister_id, DFX_DEPS).await {
+        Ok(Some(deps_raw)) => {
+            let deps_str = String::from_utf8(deps_raw)?;
+            let mut deps = vec![];
+            for entry in deps_str.split_terminator(';') {
+                match entry.split_once(':') {
+                    Some((_, p)) => {
+                        let dep_id = Principal::from_text(p)
+                            .with_context(|| format!("`{p}` is not a valid Principal."))?;
+                        canisters_to_pull.push_back(dep_id);
+                        deps.push(dep_id);
+                    }
+                    None => bail!(
+                        "Failed to parse `dfx:deps` entry: {entry}. Expected `name:Principal`. "
+                    ),
+                }
+            }
+            pulled_json.canisters.insert(
+                canister_id,
+                PulledCanister {
+                    deps,
+                    ..Default::default()
+                },
+            );
+        }
+        Ok(None) => {
+            warn!(
+                logger,
+                "`{DFX_DEPS}` metadata not found in canister {canister_id}."
+            );
+            pulled_json
+                .canisters
+                .insert(canister_id, PulledCanister::default());
+        }
+        Err(e) => {
+            bail!(e);
+        }
+    }
+    Ok(())
+}
+
+// download canister wasm, candid, init
+async fn download_canister_files(
+    env: &dyn Environment,
+    logger: &Logger,
+    canister_id: Principal,
+    pulled_canister: &mut PulledCanister,
+) -> DfxResult {
+    info!(logger, "Pulling canister {canister_id}...");
+
+    let agent = env
+        .get_agent()
+        .ok_or_else(|| anyhow!("Cannot get HTTP client from environment."))?;
+
+    // try fetch `dfx:wasm_hash`. If not available, get the hash of the on chain canister.
+    let hash_on_chain = match fetch_metatdata(agent, canister_id, DFX_WASM_HASH).await {
+        Ok(Some(wasm_hash_raw)) => {
```

Comment:
Use of `?` rather than match on Result
```suggestion
    let hash_on_chain = match fetch_metatdata(agent, canister_id, DFX_WASM_HASH).await? {
        Some(wasm_hash_raw) => {
```